sklearn/metrics/_pairwise_distances_reduction.pyx

```diff
@@ -0,0 +1,984 @@
+
+    The base class provides the generic chunked parallelization template using
+    OpenMP loops (Cython prange), either on rows of X or rows of Y depending on
+    their respective sizes.
+
+    The subclasses are specialized for reduction.
+
+    The actual distance computation for a given pair of rows of X and Y are
+    delegated to format-specific subclasses of the DatasetsPair companion base
+    class.
+
+    Parameters
+    ----------
+    datasets_pair: DatasetsPair
+        The pair of dataset to use.
+
+    chunk_size: int, default=None,
+        The number of vectors per chunk. If None (default) looks-up in
+        scikit-learn configuration for `pairwise_dist_chunk_size`,
+        and use 256 if it is not set.
+
+    n_threads: int, default=None
+        The number of OpenMP threads to use for the reduction.
+        Parallelism is done on chunks and the sharding of chunks
+        depends on the `strategy` set on :method:`~PairwiseDistancesReduction.compute`.
+
+        None and -1 means using all processors.
+    """
+
+    cdef:
+        readonly DatasetsPair datasets_pair
+
+        ITYPE_t n_threads
+        ITYPE_t effective_omp_n_thread
+        ITYPE_t n_samples_chunk, chunk_size
+
+        ITYPE_t n_samples_X, X_n_samples_chunk, X_n_chunks, X_n_samples_remainder
+        ITYPE_t n_samples_Y, Y_n_samples_chunk, Y_n_chunks, Y_n_samples_remainder
+
+    @classmethod
+    def valid_metrics(cls) -> List[str]:
+        excluded = {
+            "pyfunc",  # is relatively slow because we need to coerce data as np arrays
+            "mahalanobis", # is numerically unstable
+            # TODO: In order to support discrete distance metrics, we need to have a
+            # simultaneous sort which breaks ties on indices when distances are identical.
+            # The best might be using a std::sort and a Comparator which might need
+            # AoS instead of SoA (currently used).
+            "hamming",
+            *BOOL_METRICS,
+        }
+        return sorted({"fast_euclidean", "fast_sqeuclidean",
+                       *METRIC_MAPPING.keys()}.difference(excluded))
+
+    @classmethod
+    def is_usable_for(cls, X, Y, metric) -> bool:
+        """Return True if the PairwiseDistancesReduction can be used for the given parameters.
+
+        Parameters
+        ----------
+        X : {ndarray, sparse matrix} of shape (n_samples_X, n_features)
+            Input data.
+
+        Y : {ndarray, sparse matrix} of shape (n_samples_Y, n_features)
+            Input data.
+
+        metric : str, default='euclidean'
+            The distance metric to use.
+            For a list of available metrics, see the documentation of
+            :class:`~sklearn.metrics.DistanceMetric`.
+
+        Returns
+        -------
+        True if the PairwiseDistancesReduction can be used, else False.
+        """
+        # Coercing to np.array to get the dtype
+        # TODO: what is the best way to get lists' dtype?
```

Comment:
I would say to not handle lists and assume `{ndarray, sparse matrix}` as documented above.

Reply:
IIRC, in some cases, lists can be passed here. If it is the case, I still need to see how the validation made upstream can be adapted.

Reply:
> IIRC, in some cases, lists can be passed here.

Which cases? If there is not existing usage in scikit-learn itself, +1 for not supporting lists at all in the Cython code and let the public Python code handle input validation.

Reply:
`is_usable_for` is used in the following:

1. `pairwise_distances_argmin_min`, which validates with `check_pairwise_arrays`.
2. `pairwise_distances_argmin`, which validates with `check_pairwise_arrays`.
3. `kneighbors`, which runs it on `self._fit_X`, which was validated during `fit`.

Is there a place where `is_usable_for` is called with lists?

Reply:
> Is there a place where `is_usable_for` is called with lists?

It is called in some doctests using lists of lists, e.g.:

https://github.com/scikit-learn/scikit-learn/blob/0b8516749b2dd7211204fc8a7cb916bcac129762/sklearn/neighbors/_regression.py#L145-L152

Hence, this simple check. Still, `is_usable_for` should solely accept `{ndarray, sparse matrix}` as you said.

Reply:
`KNeighborsRegressor.fit` calls `NeighborsBase._fit` which calls `BaseEstimator._validate_data`. So `is_usable_for` should never see a direct Python list.

+1 for not doing defensive coding and removing the unecessary `asarray` checks below and instead assume that the type requirements of the `is_usable_for` docstring are always respected in our code base.

Reply:
> +1 for not doing defensive coding and removing the unecessary asarray checks below and instead assume that the type requirements of the is_usable_for docstring are always respected in our code base.

I agree, but this currently breaks code example and thus the test suite (as with the example given above for instance).

Reply:
hum I did not understand why this would happen so I tried:

```python-traceback
________________________________________________________________________ [doctest] sklearn.neighbors._base.KNeighborsMixin.kneighbors _________________________________________________________________________
680         In the following example, we construct a NearestNeighbors
681         class from an array representing our data set and ask who's
682         the closest point to [1,1,1]
683 
684         >>> samples = [[0., 0., 0.], [0., .5, 0.], [1., 1., .5]]
685         >>> from sklearn.neighbors import NearestNeighbors
686         >>> neigh = NearestNeighbors(n_neighbors=1)
687         >>> neigh.fit(samples)
688         NearestNeighbors(n_neighbors=1)
689         >>> print(neigh.kneighbors([[1., 1., 1.]]))
UNEXPECTED EXCEPTION: AttributeError("'list' object has no attribute 'dtype'")
Traceback (most recent call last):
  File "/Users/ogrisel/mambaforge/envs/dev/lib/python3.9/doctest.py", line 1336, in __run
    exec(compile(example.source, filename, "single",
  File "<doctest sklearn.neighbors._base.KNeighborsMixin.kneighbors[4]>", line 1, in <module>
  File "/Users/ogrisel/code/scikit-learn/sklearn/neighbors/_base.py", line 717, in kneighbors
    and PairwiseDistancesArgKmin.is_usable_for(
  File "sklearn/metrics/_pairwise_distances_reduction.pyx", line 167, in sklearn.metrics._pairwise_distances_reduction.PairwiseDistancesReduction.is_usable_for
    return (not issparse(X) and X.dtype == np.float64 and X.ndim == 2 and
AttributeError: 'list' object has no attribute 'dtype'
```

so the problem is in the public `kneighbors` method. I think it should be refactored to call `X = self._validate_data(X, accept_sparse="csr", reset=False)` first, then check `PairwiseDistancesArgKmin.is_usable_for` to set `use_pairwise_distances_reductions` and if True, convert to C-contiguous array with:

```python
if not sp.isparse(X) and use_pairwise_distances_reductions:
    X = np.ascontiguousarray(X)
```

Reply:
I will address it in a relevant sub-PR.